### PR TITLE
Flux changes for NXmx

### LINKED ISSUE #986: NXmx total_flux
The quantity **total_flux** is currently defined as flux incident on beam plane in _photons per second_.
For pulsed light sources, like free electron lasers, the temporal profile of each pulse is rarely reliably known.
It would be therefore desirable and more accurate to express **total_flux** in photons.

Would it be possible to have either photons/second or photons as unit of measurement for **total_flux**?

### PR DESCRIPTION
**Pull request by phyy-nx**
- Add flux_integrated and total_flux_integrated, Useful where temporal profiles of flux are not known.
- Change total_flux to not be required
- Add NXbeam attribute flux, to point to which flux field is used
- Add optional NXmonitor for additional incident beam data

Closes #986
Closes nexusformat/NIAC#96

### CODE REVIEW

**Review by yayahjb — APPROVED** [review]
Good fix -- Herbert

File: applications/NXmx.nxdl.xml
Comment (phyy-nx) [inline applications/NXmx.nxdl.xml:828]: Typo: extra per

File: applications/NXmx.nxdl.xml
Comment (phyy-nx) [inline applications/NXmx.nxdl.xml:881]: Take out maxOccurs

File: applications/NXmx.nxdl.xml
Comment (phyy-nx) [inline applications/NXmx.nxdl.xml:825]: unit should be per area (NX_PER_AREA)

File: applications/NXmx.nxdl.xml
Comment (phyy-nx) [inline applications/NXmx.nxdl.xml:836]: Units: NX_DIMENSIONLESS

**Comment by woutdenolf** [comment]
Referenced units:

- `NX_FLUX`: 1/s/cm^2
- `NX_PER_AREA`: 1/cm^2
- `NX_FREQUENCY`: 1/s
- `NX_DIMENSIONLESS`

This is what currently exists:

```
NXbeam
  flux[i]: (optional) NX_FLOAT {units=NX_FLUX}

NXmx
ENTRY: (required) NXentry
  BEAM: (required) NXbeam
   flux: (optional) NX_FLOAT {units=NX_FLUX}
   total_flux: (required) NX_FLOAT {units=NX_FREQUENCY}
```

This is the new proposal (without the `NXmonitor` part, only the *flux* related things):

```
NXbeam
  flux[i]: (optional) NX_FLOAT {units=NX_FLUX}

NXmx
ENTRY: (required) NXentry
  BEAM: (required) NXbeam
   @flux: (optional) NX_CHAR
   total_flux: (optional) NX_FLOAT {units=NX_FREQUENCY}
   flux: (optional) NX_FLOAT {units=NX_FLUX}
   flux_integrated: (optional) NX_FLOAT {units=NX_PER_AREA}
   total_flux_integrated: (optional) NX_FLOAT {units=NX_DIMENSIONLESS}
```

So the word "integrated" stands for _time integrated_ and the word "total" stands for _area integrated_.

The reason we cannot come up with sensible names like

- `flux_integrated` -> flux_time_integrated
- `total_flux` -> flux_area_integrated
- `total_flux_integrated` -> flux_time_and_area_integrated

is because `NXmx` already has `total_flux`.

Sorry @phyy-nx for my initial pessimistic view, the two units that were wrong scared the hell out of me. The names are very unfortunate but we already have `total_flux` in `NXmx` so I don't see another way either. I'd say we should not propagate these unfortunate names to `NXbeam` in the future.

I have no strong opinion on the addition of the optional `NXmonitor`.

**Comment by phyy-nx** [comment]
No worries, thanks for the comments.  I'll review your comments more carefully as soon as I get a chance.

**Comment by yayahjb** [comment]
All existing MX ontologies have some unfortunate names, but it can be a
disaster to make changes to existing tags used in large numbers of images,
in this case, millions.  The best
you can do it to start a slow process of deprecation and recommendation of
new names.
Most of the tme it is not worth the effort.

It could be worse.  We could be back to when I was young and all of physics
moved
from default left-handed systems for E&M to right handed.  I still have
trouble with that,
especially since I am a right-trained lefty.

**Comment by woutdenolf** [comment]
@phyy-nx Maybe you could add a comment about `flux` integration to the respective fields?
- `flux_integrated`: `flux` integrated over time
- `total_flux`: `flux` integrated over area
- `total_flux_integrated`: `flux` integrated over time and area

You can understand that from the units but maybe it is good to mention it explicitly since the words "total" and "integrated" are not self explanatory.

**Comment by prjemian** [comment]
@phyy-nx Since this is labeled "NIAC should review", then is there anything to do at the Code Camp this month?  If not, then remove this PR from the Code Camp project.

File: applications/NXmx.nxdl.xml
Comment (woutdenolf) [reply applications/NXmx.nxdl.xml:828]: Merge https://github.com/nexusformat/definitions/pull/1082 to fix.

File: applications/NXmx.nxdl.xml
Comment (woutdenolf) [reply applications/NXmx.nxdl.xml:825]: Merge https://github.com/nexusformat/definitions/pull/1082 to fix.

File: applications/NXmx.nxdl.xml
Comment (woutdenolf) [reply applications/NXmx.nxdl.xml:836]: Merge https://github.com/nexusformat/definitions/pull/1082 to fix.

**Comment by woutdenolf** [comment]
I proposed a fix for all my comments in Merge https://github.com/nexusformat/definitions/pull/1082. You can merge it in this branch if you approve.

**Comment by phyy-nx** [comment]
Ok, I think that's all the feedback from the spring NIAC.  I think we can remove the code camp label and vote on it this fall.

**Review by woutdenolf — APPROVED** [review]
LGTM

**Comment by prjemian** [comment]
GitHub reports a merge conflict with NXmx.nxdl.xml.

Also, since we will not merge this until the NIAC looks at it, can we convert this PR to draft state to prevent someone merging?

**Comment by phyy-nx** [comment]
Conflicts are fixed

**Comment by phyy-nx** [comment]
Conversation from code camp summer 2022. Since this needs a NIAC vote anyway, let's make as a goal to use the logic proposed in #1002 to enforce a choice between `flux`, `total_flux`, `flux_integrated`, `total_flux_integrated`, and the path to an NXmonitor with per-shot flux information.  We'd like to have that ready by the fall NIAC meeting.  Owners: @benajamin for finishing the new NXDL rule needed to enforce choices between fields and groups, and @phyy-nx to add the logic to this PR.

**Comment by phyy-nx** [comment]
From Fall 2022 NIAC, don't need to wait on #1002.  Instead, vote on this, merge it, and add a new issue to require one of `flux`, `total_flux`, `flux_integrated`, `total_flux_integrated`

**Comment by prjemian** [comment]
There is a conflict involving the NXmonitor group.

**Comment by phyy-nx** [comment]
Conflicts fixed, ready for NIAC vote

**Review by yayahjb — APPROVED** [review]

File: applications/NXmx.nxdl.xml
Comment (benajamin) [inline applications/NXmx.nxdl.xml:939]: This monitor is optional and has no recommended name or specific use case. Is it really helpful to make such a vague suggestion?

Comment (phyy-nx) [reply applications/NXmx.nxdl.xml:939]: This is optional because the user must select one of the 5 options.  It it's vague because it's a placeholder for future use.

**Comment by prjemian** [comment]
CI failed:

```
Warning, treated as error:
/home/runner/work/definitions/definitions/build/manual/source/classes/applications/NXmx.rst:1005:undefined label: 'nxmx-entry-instrument-beam-total-flux-field'
make: *** [Makefile:62: html] Error 2
Error: Process completed with exit code 2.
```

And that means my previous change was not correct.  My apology.  Please switch back to:

    :ref:`flux &lt;/NXmx/ENTRY/INSTRUMENT/BEAM/flux-field&gt;`

**Review by prjemian — APPROVED** [review]
LGTM

**Comment by benajamin** [comment]
The proposal is to accept the changes to NXmx represented by pull request #1035

**Pull request merged**
